dissect/target/plugins/os/unix/log/journal.py

```diff
@@ -0,0 +1,388 @@
+        self.entry_array_offset = self.header.entry_array_offset
+
+    def entry_object_offsets(self) -> Iterator[int]:
+        """Read object entry arrays."""
+
+        offset = self.entry_array_offset
+
+        # Entry Array with next_entry_array_offset set to 0 is the last in the list
+        while offset != 0:
+            self.fh.seek(offset)
+
+            object = c_journal.ObjectHeader(self.fh)
+
+            if object.type == c_journal.ObjectType.OBJECT_ENTRY_ARRAY:
+                # After the object is checked, read again but with EntryArrayObject instead of ObjectHeader
+                self.fh.seek(offset)
+                entry_array_object = c_journal.EntryArrayObject(self.fh)
+
+                for entry_object_offset in entry_array_object.entry_object_offsets:
+                    # Check if the offset is not zero and points to nothing
+                    if entry_object_offset:
+                        yield entry_object_offset
+
+                offset = entry_array_object.next_entry_array_offset
+
+    def decode_value(self, value: bytes) -> tuple[str, str]:
+        value = value.decode().strip()
+
+        # Remove underscores _, __
+        if value[0:1] == "__":
+            value = value[2:]
+        elif value[0] == "_":
+            value = value[1:]
+
+        key, value = value.split("=", 1)
+        key = key.lower()
+
+        return key, value
+
+    def __iter__(self) -> Iterator[Instance]:
+        "Iterate over the entry objects to read payloads."
+
+        for offset in self.entry_object_offsets():
+            self.fh.seek(offset)
+            entry = c_journal.EntryObject(self.fh)
+
+            event = {}
+            event["ts"] = ts.from_unix_us(entry.realtime)
+
+            for item in entry.items:
+                try:
+                    self.fh.seek(item.object_offset)
+
+                    object = c_journal.ObjectHeader(self.fh)
+
+                    if object.type == c_journal.ObjectType.OBJECT_DATA:
+                        self.fh.seek(item.object_offset)
+
+                        data_object = c_journal.DataObject(self.fh)
+
+                        data = data_object.payload
+
+                        if not data:
+                            # If the payload is empty
+                            continue
+                        elif data_object.flags == c_journal.ObjectFlag.OBJECT_COMPRESSED_XZ:
+                            data = lzma.decompress(data)
+                        elif data_object.flags == c_journal.ObjectFlag.OBJECT_COMPRESSED_LZ4:
+                            data = lz4.decompress(data[8:])
+                        elif data_object.flags == c_journal.ObjectFlag.OBJECT_COMPRESSED_ZSTD:
+                            data = zstandard.decompress(data)
+
+                        key, value = self.decode_value(data)
+                        event[key] = value
+
+                except Exception:
+                    continue
```

Comment:
Just a question...

Not really a systemd-journal expert but I see this Exception here is to catch invalid offsets, as in... (sample):

```
<EntryItem object_offset=0x3905a0, hash=0x64fb679e5714f4d>
3736992
<EntryItem object_offset=0x390640, hash=0x6fbd0df70f2eae8c>
3737152
<EntryItem object_offset=0x3906e8, hash=0x180e145321813b2b>
3737320
<EntryItem object_offset=0x6, hash=0x38>
6
<EntryItem object_offset=0x391a58, hash=0x390780>
3742296
<EntryItem object_offset=0x390cf0, hash=0x390f28>
3738864
<EntryItem object_offset=0x391648, hash=0x1>
3741256
<EntryItem object_offset=0x67, hash=0x4a45e1aeed5a0fb2>
103
<EntryItem object_offset=0x0, hash=0x0>
0
<EntryItem object_offset=0x390cf0, hash=0x0>
3738864
<EntryItem object_offset=0x1, hash=0x5f454352554f535f>
1
<EntryItem object_offset=0x494e4f544f4e4f4d, hash=0x5453454d49545f43>
5282246636478549837
[Errno 22] Invalid argument
```

The last one seems a negative number at first glance.. Is this normal? I could not find anything about this in the systemd docs, it just says all offsets are 64bit unsigned integers in LE format (https://www.freedesktop.org/wiki/Software/systemd/journal-files/).

Reply:
Might have something to do with HEADER_INCOMPATIBLE_COMPACT?

Reply:
After all entries is a union... https://systemd.io/JOURNAL_FILE_FORMAT/

Reply:
Some quick research...

```
struct EntryItemCompact {
    le32_t object_offset;
    le32_t nope1;
    le64_t nope2;
};

struct EntryObject {
    ....
    EntryItemCompact items[size / 64];
};
```

seems to be working for compact entries for me (no more exceptions).

Reply:
For me also no exceptions on two different system, I left the `Exception` in the code in the case the format is updated which could result in errors.

Reply:
Can you add logging to this exception? It would be nice for everyone to see if anything goes wrong.